samza-azure/src/main/java/org/apache/samza/system/eventhub/producer/EventHubSystemProducer.java

```diff
@@ -0,0 +1,337 @@
+        aggEventByteWriteRate = registry.newCounter(AGGREGATE, EVENT_BYTE_WRITE_RATE);
+        aggSendLatency = new SamzaHistogram(registry, AGGREGATE, SEND_LATENCY);
+        aggSendCallbackLatency = new SamzaHistogram(registry, AGGREGATE, SEND_CALLBACK_LATENCY);
+        aggSendErrors = registry.newCounter(AGGREGATE, SEND_ERRORS);
+      }
+    }
+
+    isStarted = true;
+  }
+
+  @Override
+  public synchronized void stop() {
+    LOG.info("Stopping event hub system producer...");
+    streamPartitionSenders.values().forEach((streamPartitionSender) -> {
+        List<CompletableFuture<Void>> futures = new ArrayList<>();
+        streamPartitionSender.forEach((key, value) -> futures.add(value.close()));
+        CompletableFuture<Void> future = CompletableFuture.allOf(futures.toArray(new CompletableFuture[futures.size()]));
+        try {
+          future.get(config.getShutdownWaitTimeMS(), TimeUnit.MILLISECONDS);
+        } catch (ExecutionException | InterruptedException e) {
+          LOG.warn("Closing the partition sender failed ", e);
+        } catch (TimeoutException e) {
+          LOG.warn("Closing the partition sender timed out ", e);
+        }
+      });
+    eventHubClients.values().forEach(ehClient -> ehClient.close(config.getShutdownWaitTimeMS()));
+    eventHubClients.clear();
+  }
+
+  @Override
+  public synchronized void register(String streamName) {
+    LOG.info("Trying to register {}.", streamName);
+    if (isStarted) {
+      String msg = "Cannot register once the producer is started.";
+      LOG.error(msg);
+      throw new SamzaException(msg);
+    }
+
+    EventHubClientWrapper ehClient = eventHubClientWrapperFactory
+            .getEventHubClientWrapper(config.getStreamNamespace(streamName), config.getStreamEntityPath(streamName),
+                    config.getStreamSasKeyName(streamName), config.getStreamSasToken(streamName), config);
+
+    ehClient.init();
+    eventHubClients.put(streamName, ehClient);
+    streamPartitionSenders.put(streamName, new HashMap<>());
+  }
+
+  @Override
+  public synchronized void send(String destination, OutgoingMessageEnvelope envelope) {
+    if (!isStarted) {
+      throw new SamzaException("Trying to call send before the producer is started.");
+    }
+
+    if (!eventHubClients.containsKey(destination)) {
+      String msg = String.format("Trying to send event to a destination {%s} that is not registered.", destination);
+      LOG.error(msg);
+      throw new SamzaException(msg);
+    }
+
+    if (sendExceptionOnCallback != null) {
+      SamzaException e = new SamzaException(sendExceptionOnCallback);
+      sendExceptionOnCallback = null;
+      pendingFutures.clear();
+      LOG.error("One of the previous sends failed.");
+      throw e;
+    }
+
+    EventData eventData = createEventData(destination, envelope);
+    int eventDataLength =  eventData.getBytes() == null ? 0 : eventData.getBytes().length;
+    eventWriteRate.get(destination).inc();
```

Comment:
@vjagadish1989 would be good to check that these metrics aren't duplicating those that samza emits anyway.